feat: add delegate routers to libp2p config

jacobheun opened the pull request:
Adds the delegate routing modules needed for https://github.com/ipfs/js-ipfs/issues/2155.

This enables js-ipfs nodes to leverage external nodes to provide content, find content, and find peers.

I added a crude randomizer for the preload nodes which are supporting delegate routing. Ideally this would be per request, but the modules will need to be updated to take multiple hosts to do so.

**A note**: If the DHT is enabled on your node, libp2p will attempt to perform the action using the DHT first. If that is not successful, it will then use the delegates to perform the request. This gives nodes the ability to fallback to the external nodes while they are building up their DHT records/peers.

jacobheun commented:
> Should the hostnames be in config?

Yeah, we should probably have the multiaddrs in there and then pull out the http endpoints.

This will cause the config to deviate from go, as we'll likely add a new address field like `Addresses.Delegates`. This should be fine as it's more of a js thing, but if we do that we should check to make sure go doesn't care about the extra field.

Does `Addresses.Delegates` sound like the right place for this to go?
```js
Addresses: {
  ...
  Delegates: [
    '/dns4/node0.preload.ipfs.io/tcp/443/https',
    '/dns4/node1.preload.ipfs.io/tcp/443/https'
  ]
}
```

dirkmc commented:
`Discovery` and `Bootstrap` are at the top level, so it may make sense for Delegates to be as well, and thereby less likely to conflict with go config

jacobheun commented:
I've updated this to only add delegated routing if the delegates multiaddrs are actually added to the Config `Addresses.Delegates` property. I also added docs and a couple tests.

alanshaw commented:
@jacobheun I fixed the link and rebased
